app/scripts/metamask-controller.js

```diff
@@ -5748,6 +5755,154 @@ export default class MetamaskController extends EventEmitter {
+    const shouldShowApproval = permittedAccountStatus === undefined;
+
+    // We should redirect to the referral url if the account is approved
+    const shouldRedirect = permittedAccountStatus === ReferralStatus.Approved;
+
+    // If we shouldn't show the approval screen and shouldn't redirect, return
+    if (!shouldShowApproval && !shouldRedirect) {
+      return;
+    }
+
+    if (shouldShowApproval) {
+      // method will fail if there is an existing approval request
+      // or if the user rejects the request (e.g. by disconnecting)
+      const approvalResponse = await this.approvalController.add({
+        origin,
+        type: HYPERLIQUID_APPROVAL_TYPE,
+        requestData: { selectedAddress: permittedAccount },
+        shouldShowRequest:
+          triggerType !==
+          HyperliquidPermissionTriggerType.OnNavigateConnectedTab,
+      });
+
+      if (approvalResponse?.approved) {
+        this._handleHyperliquidApprovedAccount(
+          permittedAccount,
+          permittedAccounts,
+          declinedAccounts,
+        );
+        await this._handleHyperliquidReferralRedirect(
+          tabId,
+          HYPERLIQUID_ORIGIN,
+          permittedAccount,
+        );
+      } else {
+        this.preferencesController.addReferralDeclinedAccount(permittedAccount);
+      }
+    }
+
+    if (shouldRedirect) {
+      await this._handleHyperliquidReferralRedirect(
+        tabId,
+        HYPERLIQUID_ORIGIN,
+        permittedAccount,
+      );
+    }
+  }
+
+  /**
+   * Handles redirection to the Hyperliquid referral page.
+   *
+   * @param {number} tabId - The browser tab ID to update.
+   * @param {string} hyperliquidOrigin - The Hyperliquid origin URL.
+   * @param {string} permittedAccount - The permitted account.
+   */
+  async _handleHyperliquidReferralRedirect(
+    tabId,
+    hyperliquidOrigin,
+    permittedAccount,
+  ) {
+    await this._updateHyperliquidReferralUrl(tabId, hyperliquidOrigin);
+    // Mark this account as having been shown the Hyperliquid referral page
+    this.preferencesController.addReferralPassedAccount(permittedAccount);
+  }
+
+  /**
+   * Handles referral states for permitted accounts after user approval.
+   *
+   * @param {string} permittedAccount - The permitted account.
+   * @param {string[]} permittedAccounts - The permitted accounts.
+   * @param {string[]} declinedAccounts - The previously declined permitted accounts.
+   */
+  _handleHyperliquidApprovedAccount(
+    permittedAccount,
+    permittedAccounts,
+    declinedAccounts,
+  ) {
+    if (declinedAccounts.length === 0) {
+      // If there are no previously declined permitted accounts then
+      // we approve all permitted accounts so that the user is not
+      // shown the approval screen unnecessarily when switching
+      this.preferencesController.setAllAccountsReferralApproved(
+        permittedAccounts,
+      );
+    } else {
+      this.preferencesController.addReferralApprovedAccount(permittedAccount);
```

Comment:
thoughts on dropping permittedAccount and just using `permittedAccounts[0]` instead?

The other less fragile way to approach this would be to not do the `.slice(1)` below. `permittedAccount` should never be in `declinedAccounts` anyway